libproxy: update to 0.5.9; add new option variants

lcvisser opened the pull request:
#### Description

The libproxy implementation was rewritten with 0.5.0 and switched to meson as build system. The new build options do not match 1-to-1 to the old options and all language bindings were removed in favor of gobject-introspection.

* Update to libproxy 0.5.9
* Use meson instead of CMake
* Add variants for new options

Closes: https://trac.macports.org/ticket/67730
Closes: https://trac.macports.org/ticket/71389

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 15.1.1 24B91 x86_64
Command Line Tools 16.1.0.0.1.1729049160

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

mascguy commented on net/libproxy/Portfile:
```diff
@@ -2,11 +2,11 @@
-github.setup        libproxy libproxy 0.4.18
-revision            2
+github.setup        libproxy libproxy 0.5.9
+revision            3
```
Since you're upgrading to a new version, reset `revision` to zero.

mascguy replied:
Fixed

barracuda156 commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
-depends_lib-append  path:bin/vala:vala \
-                    path:lib/pkgconfig/glib-2.0.pc:glib2
+depends_lib-append  \
+                    port:icu \
```
There is an `icu-devel` port, so this should be path-style dependency.

mascguy replied:
Fixed

barracuda156 commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
+depends_lib-append  \
+                    port:icu \
+                    port:gsettings-desktop-schemas \
+                    port:gobject-introspection
```
And this as well.

ryandesign replied:
Ports that use gobject introspection typically do so by including the gobject_introspection-1.0 portgroup to handle various common problems, however I don't know if the portgroup has been updated for compatibility with meson build systems. If it hasn't, it would be a good idea to update it, and then include it in this port.

barracuda156 replied:
It may be too much of a burden to put onto the PR author. This port builds fine without that PG, so probably PG can be handled separately (there are a lot of ports depending on GO without adding its PG, including meson-using ones).

barracuda156 commented:
It also fails to configure, not finding a dependency:
```
--->  Extracting libproxy-0.5.9.tar.gz
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work" && /opt/x86_64/bin/gzip -dc '/opt/local/var/macports/distfiles/libproxy/libproxy-0.5.9.tar.gz' | /opt/x86_64/bin/gnutar --no-same-owner -xf - 
--->  Configuring libproxy
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/libproxy-0.5.9" && /opt/local/bin/meson setup --prefix=/opt/local -Ddocs=false -Dtests=false -Dconfig-env=false -Dconfig-gnome=true -Dconfig-kde=true -Dconfig-windows=false -Dconfig-sysconfig=false -Dconfig-osx=false -Dconfig-xdp=false -Dpacrunner-duktape=true -Dvapi=false -Dcurl=false -Dintrospection=true -Drelease=true /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/libproxy-0.5.9 /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/build --wrap-mode=default 
The Meson build system
Version: 1.5.1
Source dir: /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/libproxy-0.5.9
Build dir: /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/build
Build type: native build
Project name: libproxy
Project version: 0.5.9
C compiler for the host machine: /opt/local/bin/gcc-mp-14 (gcc 14.2.0 "gcc-mp-14 (MacPorts gcc14 14.2.0_1+stdlib_flag) 14.2.0")
C linker for the host machine: /opt/local/bin/gcc-mp-14 ld64 97.17
Host machine cpu family: ppc
Host machine cpu: power macintosh
Compiler for C supports arguments -Wcast-align: YES 
Compiler for C supports arguments -Wdeclaration-after-statement: YES 
Compiler for C supports arguments -Werror=address: YES 
Compiler for C supports arguments -Werror=array-bounds: YES 
Compiler for C supports arguments -Werror=empty-body: YES 
Compiler for C supports arguments -Werror=implicit: YES 
Compiler for C supports arguments -Werror=implicit-function-declaration: YES 
Compiler for C supports arguments -Werror=incompatible-pointer-types: YES 
Compiler for C supports arguments -Werror=init-self: YES 
Compiler for C supports arguments -Werror=int-conversion: YES 
Compiler for C supports arguments -Werror=int-to-pointer-cast: YES 
Compiler for C supports arguments -Werror=main: YES 
Compiler for C supports arguments -Werror=misleading-indentation: YES 
Compiler for C supports arguments -Werror=missing-braces: YES 
Compiler for C supports arguments -Werror=missing-include-dirs: YES 
Compiler for C supports arguments -Werror=nonnull: YES 
Compiler for C supports arguments -Werror=overflow: YES 
Compiler for C supports arguments -Werror=parenthesis: NO 
Compiler for C supports arguments -Werror=pointer-arith: YES 
Compiler for C supports arguments -Werror=pointer-to-int-cast: YES 
Compiler for C supports arguments -Werror=redundant-decls: YES 
Compiler for C supports arguments -Werror=return-type: YES 
Compiler for C supports arguments -Werror=sequence-point: YES 
Compiler for C supports arguments -Werror=shadow: YES 
Compiler for C supports arguments -Werror=strict-prototypes: YES 
Compiler for C supports arguments -Werror=trigraphs: YES 
Compiler for C supports arguments -Werror=undef: YES 
Compiler for C supports arguments -Werror=write-strings: YES 
Compiler for C supports arguments -Wformat-nonliteral: YES 
Compiler for C supports arguments -Wignored-qualifiers: YES 
Compiler for C supports arguments -Wimplicit-function-declaration: YES 
Compiler for C supports arguments -Wlogical-op: YES 
Compiler for C supports arguments -Wmissing-declarations: YES 
Compiler for C supports arguments -Wmissing-format-attribute: YES 
Compiler for C supports arguments -Wmissing-include-dirs: YES 
Compiler for C supports arguments -Wmissing-noreturn: YES 
Compiler for C supports arguments -Wnested-externs: YES 
Compiler for C supports arguments -Wno-cast-function-type: YES 
Compiler for C supports arguments -Wno-dangling-pointer: YES 
Compiler for C supports arguments -Wno-missing-field-initializers: YES 
Compiler for C supports arguments -Wno-sign-compare: YES 
Compiler for C supports arguments -Wno-unused-parameter: YES 
Compiler for C supports arguments -Wold-style-definition: YES 
Compiler for C supports arguments -Wpointer-arith: YES 
Compiler for C supports arguments -Wredundant-decls: YES 
Compiler for C supports arguments -Wstrict-prototypes: YES 
Compiler for C supports arguments -Wswitch-default: YES 
Compiler for C supports arguments -Wswitch-enum: YES 
Compiler for C supports arguments -Wundef: YES 
Compiler for C supports arguments -Wuninitialized: YES 
Compiler for C supports arguments -Wunused: YES 
Compiler for C supports arguments -fno-strict-aliasing: YES 
Compiler for C supports arguments -Werror=format-security -Werror=format=2: YES 
Compiler for C supports link arguments -Wl,-z,nodelete: NO 
Found pkg-config: YES (/opt/local/bin/pkg-config) 0.29.2
Run-time dependency glib-2.0 found: YES 2.78.4
Run-time dependency gio-2.0 found: YES 2.78.4
Run-time dependency libcurl found: YES 8.10.1
Library ws2_32 found: NO
Run-time dependency gsettings-desktop-schemas found: YES 3.38.0
Configuring px-backend-config.h using configuration
Found CMake: /opt/local/bin/cmake (3.28.4)
Run-time dependency duktape found: NO (tried pkgconfig, framework and cmake)

src/backend/plugins/pacrunner-duktape/meson.build:5:14: ERROR: Dependency "duktape" not found, tried pkgconfig, framework and cmake

A full log can be found at /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_libproxy/libproxy/work/build/meson-logs/meson-log.txt
```

barracuda156 commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
-    depends_lib-append port:python311
+variant vapi description {Build with support for vapi} {
+    depends_lib-append port:vala
     configure.args-append \
```
All these should be `configure.args-replace`, because default configure args above have `-Dvapi=false` (and others), and variant needs to change those, not  add a contradicting argument.

And path-style `vala`.

So, for example:
```
depends_build-append path:bin/vala:vala
configure.args-replace -Dvapi=false -Dvapi=true
```
(I think Vala is a build dependency, please check that.)

mascguy replied:
All fixed

lcvisser replied:
I think you're right re: build dependency. It appears that enabling the option only results in a call to generate some auxiliary VAPI file

- https://github.com/libproxy/libproxy/blob/main/src/libproxy/meson.build#L82
- https://mesonbuild.com/Gnome-module.html#gnomegenerate_vapi

barracuda156 reviewed (changes requested):
Besides minor issues, which need to be addressed, are there any reason not to enable `vala` by default? While I honestly do not know what benefit it gives, I encountered a situation when a dependent required a port to be built with Vala, otherwise it failed.

ryandesign commented:
> are there any reason not to enable `vala` by default?

See https://trac.macports.org/ticket/71389 for the request not to depend on vala by default in another port; the reasoning may apply here as well.

ryandesign commented:
> * [x]  followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?

Per those guidelines, the commit message should start with the port name and a colon and a space.

ryandesign commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
-#
-# TODO: add variants that safely enable these features
-
 use_parallel_build no
```
Since this is now using a completely different build system, it seems likely that whatever was preventing the parallel build from working before isn't anymore; can you try if it still builds after removing this?

mascguy replied:
Fixed

ryandesign commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
-    set python_prefix ${frameworks_dir}/Python.framework/Versions/3.10
-    depends_lib-append port:python310
+variant duktape description {Build with support for PAC Runner Duktape} {
+    depends_lib-append port:duktape
```
The port's `configure.args` contain `-Dpacrunner-duktape=true`. Maybe that was supposed to be `-Dpacrunner-duktape=false`, and then this variant should `…-replace` it with `…=true`?

mascguy replied:
Fixed

ryandesign commented on net/libproxy/Portfile:
```diff
@@ -18,94 +18,53 @@ long_description    Libproxy exists to answer the question: Given \
-    set python_prefix ${frameworks_dir}/Python.framework/Versions/3.11
-    depends_lib-append port:python311
+variant vapi description {Build with support for vapi} {
+    depends_lib-append port:vala
```
Because there is a vala-devel port, this dependency should be written in `path:`-style like the other ports that depend on vala do.

barracuda156 commented:
> > are there any reason not to enable `vala` by default?
> 
> 
> 
> See https://trac.macports.org/ticket/71389 for the reasoning not to depend on vala by default.
> 
> 
> 
> Because of the need to depend on it in other ports, it should be a subport, not a variant.

This should be changed across a lot of ports then, since now it is a variant (if not everywhere then at least with many ports).

Re need to depend: that was my argument to enable the variant by default. If it is just enabled everywhere, everything will work without extra subports.

mascguy commented:
> > * [x]   followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
> 
> Per those guidelines, the commit message should start with the port name and a colon and a space.

Fixed

lcvisser commented:
Thanks @mascguy for all the fixes. My day job was a bit busy today, couldn't find time to jump in.

> > > are there any reason not to enable `vala` by default?
> > 
> > 
> > See https://trac.macports.org/ticket/71389 for the reasoning not to depend on vala by default.
> > Because of the need to depend on it in other ports, it should be a subport, not a variant.
> 
> This should be changed across a lot of ports then, since now it is a variant (if not everywhere then at least with many ports).
> 
> Re need to depend: that was my argument to enable the variant by default. If it is just enabled everywhere, everything will work without extra subports.

I didn't find any ports that depend on `libproxy` and its `vala` bindings: none of the 7 ports that depend on`libproxy` have a dependency on `vala`. Hence, I concluded that making `vapi` a variant was safe and wouldn't break anything. Wrong conclusion?

I'd be happy to take a stab at making a sub-port, if that is a more desirable way forward.

lcvisser commented:
> Thanks @mascguy for all the fixes. My day job was a bit busy today, couldn't find time to jump in.
> 
> > > > are there any reason not to enable `vala` by default?
> > > 
> > > 
> > > See https://trac.macports.org/ticket/71389 for the reasoning not to depend on vala by default.
> > > Because of the need to depend on it in other ports, it should be a subport, not a variant.
> > 
> > 
> > This should be changed across a lot of ports then, since now it is a variant (if not everywhere then at least with many ports).
> > Re need to depend: that was my argument to enable the variant by default. If it is just enabled everywhere, everything will work without extra subports.
> 
> I didn't find any ports that depend on `libproxy` and its `vala` bindings: none of the 7 ports that depend on`libproxy` have a dependency on `vala`. Hence, I concluded that making `vapi` a variant was safe and wouldn't break anything. Wrong conclusion?
> 
> I'd be happy to take a stab at making a sub-port, if that is a more desirable way forward.

I'm confused at what the consensus is, if any... Sub-port or keeping the variant?

mascguy commented:
> I'm confused at what the consensus is, if any... Sub-port or keeping the variant?

It's not clear whether any dependents actually need the VAPI functionality. Have you tried building the various dependents locally, with your updated port for `libproxy`?

lcvisser commented:
> > I'm confused at what the consensus is, if any... Sub-port or keeping the variant?
> 
> It's not clear whether any dependents actually need the VAPI functionality. 

None of the dependents have `vala` as a dependency, as I mentioned. I don't see how they could use the VAPI without having `vala`, but admittedly I don't know much about `vala`.

> Have you tried building the various dependents locally, with your updated port for `libproxy`?

I only tried `wget`, which appears to work fine. Builds without errors, at least. However, having an installed "normal" ports and then installing the custom one from the repo, I'm not entirely sure I got it right. `port installed` gives `libproxy @0.5.9_0 (active)`, so I guess I did it right, but I'm not sure how I can verify that `wget` was built against that version too.

Dave-Allured commented:
@barracuda156 What situation required a port to be built with Vala?  Earlier you said:

> ... I encountered a situation when a dependent required a port to be built with Vala, otherwise it failed.

lcvisser commented:
> > Thanks @mascguy for all the fixes. My day job was a bit busy today, couldn't find time to jump in.
> > > > > are there any reason not to enable `vala` by default?
> > > > 
> > > > 
> > > > See https://trac.macports.org/ticket/71389 for the reasoning not to depend on vala by default.
> > > > Because of the need to depend on it in other ports, it should be a subport, not a variant.
> > > 
> > > 
> > > This should be changed across a lot of ports then, since now it is a variant (if not everywhere then at least with many ports).
> > > Re need to depend: that was my argument to enable the variant by default. If it is just enabled everywhere, everything will work without extra subports.
> > 
> > 
> > I didn't find any ports that depend on `libproxy` and its `vala` bindings: none of the 7 ports that depend on`libproxy` have a dependency on `vala`. Hence, I concluded that making `vapi` a variant was safe and wouldn't break anything. Wrong conclusion?
> > I'd be happy to take a stab at making a sub-port, if that is a more desirable way forward.
> 
> I'm confused at what the consensus is, if any... Sub-port or keeping the variant?

There was a [comment](https://trac.macports.org/ticket/71389#comment:12) from @ryandesign  on the ticket in Trac that I somehow overlooked but makes total sense. The fact that there are _currently_ no port ports depending on `libproxy` and `vala` doesn't mean there won't be one in the future, and that future port can only depend on sub-ports, not on variants.

I'll go ahead and try a sub-port for the `vapi` option.

Dave-Allured commented:
> ... I'll go ahead and try a sub-port for the `vapi` option.

Yes that all makes sense.  However, subports are tricky.  I have not yet figured them out myself.  If you would like to shortcut, then push what you have now with variant `vala`.  Then make the subport later.  If something breaks, then we will have learned something.  ;-)

It's just a suggestion.  Do what works best for you.

mascguy commented:
> I'll go ahead and try a sub-port for the `vapi` option.

I pushed changes to add a subport, to save you time.

Dave-Allured commented:
Nice.  Thanks @mascguy @lcvisser @barracuda156 @ryandesign !

mascguy commented:
> Why was icu added to this? Doesn't seem to depend on it.

Good catch, thanks

lcvisser commented:
`icu` is listed as required dependency [here](https://libproxy.github.io/libproxy/build-steps.html#dependencies_1)

mohd-akram commented:
@lcvisser Thanks. Seems like it's a mistake in the documentation since it's not mentioned for the other platforms (or maybe it's just a Homebrew thing).

lcvisser commented:
> @lcvisser Thanks. Seems like it's a mistake in the documentation since it's not mentioned for the other platforms (or maybe it's just a Homebrew thing).

I hadn't noticed it was not in the other platforms! I don't recall if I added it "blindly" by following the list in the documentation, or if I added it after a building failed without it...

lcvisser commented:
> > @lcvisser Thanks. Seems like it's a mistake in the documentation since it's not mentioned for the other platforms (or maybe it's just a Homebrew thing).
> 
> I hadn't noticed it was not in the other platforms! I don't recall if I added it "blindly" by following the list in the documentation, or if I added it after a building failed without it...

Apparently the first... I just tried without it, and it builds without issue. Sorry!
